Nested jars thread safety

ilbinek opened the pull request:
Multiple JRD instances can be run at the same time and use nested jars.
Editing one nested jar from more different JRDs is not supported, this only ensures that temp files are unique for each JRD instance.

Fixes #116

judovana commented:
I disagree.


The folder shoudl be shared between isntances, as it was. Only the jars/folders should not be IDed by name, but by path. If two JRDs are editing the SAME jar, they both have to see updated version. If they edit different jar, but with same name, of course they need to see state of THEIRS state. That is the only issue here.

ilbinek commented:
JRD uses those temp file mainly for reading and editing and does not extract more, than it needs.  If another JRD is running and using Local FS functions, the cache created by the second JRD will overwrite the cache of the first JRD. That's what this fix fixes.

As the cache was not designed for two JRDs to be editing the same file, it checked whether current class was already unpacked and if so, it didn't unpack it again and just returned the already unpacked one in cache. This one does not contain anything that was changed in the original. a856b0d fixes this issue and always refreshes the cache after clicking on the class in the list.

This also fixes the issue that AurumTheEnd encountered. Editing two files with the same nested jar will no longer confuse the cache and it will work as expected.

I believe this solution is the safest, but we are loosing a little bit of performance, as we are always re caching the file, even though it is already unpacked. This could be solved by adding a special button that will force the cache to be deleted and unpacked again.

Just to remind you, updated version is not stored in the cache, it's always propagated to the original jar and cache never holds anything new. Once cache is edited, it's packed and updated into the original.

judovana commented:
> JRD uses those temp file mainly for reading and editing and does not extract more, than it needs. If another JRD is running and using Local FS functions, the cache created by the second JRD will overwrite the cache of the first JRD. That's what this fix fixes.
> 

How it can override it, if it already read from modified jar?

> As the cache was not designed for two JRDs to be editing the same file, it checked whether current class was already unpacked and if so, it didn't unpack it again and just returned the already unpacked one in cache. 

Sounds correct. That is same effect, if two JRDs connect single running vm. 

> This one does not contain anything that was changed in the original. [a856b0d](https://github.com/pmikova/java-runtime-decompiler/commit/a856b0d8cfea0c75a8d6f4596ae597f31afac703) fixes this issue and always refreshes the cache after clicking on the class in the list.
> 
> This also fixes the issue that AurumTheEnd encountered. Editing two files with the same nested jar will no longer confuse the cache and it will work as expected.

The only issue here is,  that if jars of same name  are stored in different locatin, nested jars judge by its name, that it is the same jar. that is wrong. Cache must use full path to unpack the jar, to use full path as ID.
> 
> I believe this solution is the safest, but we are loosing a little bit of performance, as we are always re caching the file, even 

No. If the identical jar is edited (identical == same full path, not just name), then two jrds should always see the most current version.

> though it is already unpacked. This could be solved by adding a special button that will force the cache to be deleted and unpacked again.
> 
> Just to remind you, updated version is not stored in the cache, it's always propagated to the original jar and cache never holds anything new. Once cache is edited, it's packed and updated into the original.

That may change my view on things. Prokp Tunel will know more.

judovana commented:
Ok. Had small Chat with Ondra.
Looks like random prefix per JRD is indeed necessary.  In addition it need to include in  ID full path of each unpacked resource - not only name  - and ID of FS VM enforcing this upack. Thaxn for patience!

ilbinek commented:
> Ok. Had small Chat with Ondra.
> Looks like random prefix per JRD is indeed necessary. In addition it need to include in ID full path of each unpacked resource - not only name - and ID of FS VM enforcing this upack. Thaxn for patience!

I don't believe this is necessary. The file which is being worked on is always passed from FSAgent, ArchiveManager doesn't care if the file changed, or didn't. With current implementation, it will perform scan and unpack, if necessary, always. I believe desired behavior is already achieved.

ilbinek commented:
> > This one does not contain anything that was changed in the original. [a856b0d](https://github.com/pmikova/java-runtime-decompiler/commit/a856b0d8cfea0c75a8d6f4596ae597f31afac703) fixes this issue and always refreshes the cache after clicking on the class in the list.
> > This also fixes the issue that AurumTheEnd encountered. Editing two files with the same nested jar will no longer confuse the cache and it will work as expected.
> 
> The only issue here is, that if jars of same name are stored in different locatin, nested jars judge by its name, that it is the same jar. that is wrong. Cache must use full path to unpack the jar, to use full path as ID.
> 
I agree, this was made as a optimization feature. Adding a full path to the jar being worked on would fix the issue too, but the issue was fixed by always refreshing it and omitting the check at all, as it's necessary if two JRDs are working on the same class in the same jar.

> > I believe this solution is the safest, but we are loosing a little bit of performance, as we are always re caching the file, even
> 
> No. If the identical jar is edited (identical == same full path, not just name), then two jrds should always see the most current version.
> 
They do, if you click on the class again, it will get the most current version.
> > though it is already unpacked. This could be solved by adding a special button that will force the cache to be deleted and unpacked again.
> > Just to remind you, updated version is not stored in the cache, it's always propagated to the original jar and cache never holds anything new. Once cache is edited, it's packed and updated into the original.
> 
> That may change my view on things. Prokp Tunel will know more.

I had a chat with Prokop Tunel yesterday and I believe we agreed on current state to be the best one.

judovana commented:
Sorry. To include full path and VM ID is mandtory. Dot.

AurumTheEnd commented:
@ilbinek 
> I don't believe this is necessary.

Would it be a problem to implement though? Creating an the ID from FS VM classpath & FS VM PID? Trying to still reach a middle ground.

ilbinek commented:
> @ilbinek
> 
> > I don't believe this is necessary.
> 
> Would it be a problem to implement though? Creating an the ID from FS VM classpath & FS VM PID? Trying to still reach a middle ground.

No, I'll add it. I just don't see the purpose now, as I am not sure how and for what to use them.

AurumTheEnd commented:
@judovana forgot to close it.
